packages/components/src/number-control/index.tsx

```diff
@@ -35,32 +38,36 @@ export function NumberControl(
 		type: typeProp = 'number',
 		value: valueProp,
 		...props
-	},
-	ref
+	}: WordPressComponentProps< NumberControlProps, 'input', false >,
+	ref: ForwardedRef< any >
 ) {
 	const isStepAny = step === 'any';
+	// @ts-expect-error step should be a number but could be string
 	const baseStep = isStepAny ? 1 : parseFloat( step );
 	const baseValue = roundClamp( 0, min, max, baseStep );
-	const constrainValue = ( value, stepOverride ) => {
+	const constrainValue = ( value: number, stepOverride?: number | null ) => {
 		// When step is "any" clamp the value, otherwise round and clamp it.
 		return isStepAny
 			? Math.min( max, Math.max( min, value ) )
 			: roundClamp( value, min, max, stepOverride ?? baseStep );
 	};
 
-	const autoComplete = typeProp === 'number' ? 'off' : null;
+	const autoComplete = typeProp === 'number' ? 'off' : undefined;
 	const classes = classNames( 'components-number-control', className );
 
 	/**
 	 * "Middleware" function that intercepts updates from InputControl.
 	 * This allows us to tap into actions to transform the (next) state for
 	 * InputControl.
 	 *
-	 * @param {Object} state  State from InputControl
-	 * @param {Object} action Action triggering state change
-	 * @return {Object} The updated state to apply to InputControl
+	 * @return The updated state to apply to InputControl
 	 */
-	const numberControlStateReducer = ( state, action ) => {
+	const numberControlStateReducer = (
+		/** State from InputControl. */
+		state: InputState,
+		/** Action triggering state change. */
+		action: inputControlActionTypes.InputAction
+	) => {
```

Comment:
We could also type `numberControlStateReducer` as `NumberControlProps[ '__unstableStateReducer' ]`, which would remove the need to type `state` and `action` individually

Reply:
Makes sense 👍